### PR TITLE
[FIRRTL] Add an emitter for options and instance choices

### PR DESCRIPTION
**Pull request by nandor**

### CODE REVIEW

**Comment by dtzSiFive** [comment]
(When ready, please add test for this in `emit-basic.mlir` or so)

File: test/Dialect/FIRRTL/emit-basic.mlir [hunk 1]
```diff
@@ -744,4 +744,28 @@ firrtl.circuit "Foo" {
...
+  firrtl.extmodule private @ASICTarget()
+
+  // CHECK-LABEL: module InstChoice :
+  firrtl.module public @InstChoice() {
+    // CHECK: instchoice inst of DefaultTarget, Platform :
+    // CHECK-NEXT:    FPGA => FPGATarget
+    // CHECK-NEXT:    ASIC => ASICTarget
+    firrtl.instance_choice inst @DefaultTarget alternatives @Platform
+      { @FPGA -> @FPGATarget, @ASIC -> @ASICTarget } ()
+  }
```
Comment (seldridge) [inline test/Dialect/FIRRTL/emit-basic.mlir:770]: Nice test!

File: lib/Dialect/FIRRTL/Export/FIREmitter.cpp [hunk 2]
```diff
@@ -803,6 +821,33 @@ void Emitter::emitStatement(InstanceOp op) {
...
+
+  SmallString<16> portName(legalName);
+  portName.push_back('.');
+  unsigned baseLen = portName.size();
+  for (unsigned i = 0, e = op.getNumResults(); i < e; ++i) {
+    portName.append(legalize(op.getPortName(i)));
+    addValueName(op.getResult(i), portName);
+    portName.resize(baseLen);
+  }
+}
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Export/FIREmitter.cpp:849]: This looks good. Thanks for remembering to pass this through legalization. This should enable numeric identifiers here which is, admittedly odd, but consistent with everything else.

File: lib/Dialect/FIRRTL/Export/FIREmitter.cpp [hunk 3]
```diff
@@ -531,6 +534,21 @@ void Emitter::emitDeclaration(LayerOp op) {
   });
 }
 
+/// Emit an option declaration.
+void Emitter::emitDeclaration(OptionOp op) {
+  startStatement();
+  ps << "option " << PPExtString(op.getSymName()) << " :";
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Export/FIREmitter.cpp:540]: I think this is missing a legalize:

```suggestion
  ps << "option " << PPExtString(legalize(op.getSymName())) << " :";
```

**Review by seldridge — APPROVED** [review]

**Pull request merged**